### PR TITLE
Optionally limit concurrency of job execution

### PR DESCRIPTION
**Pull request by patrickmann**
References Graylog2/graylog-plugin-enterprise#6313
/prd Graylog2/graylog-plugin-enterprise#6399

Adds the ability to configure clusterwide max concurrency for specified job types. No more than this number of worker
threads will be scheduled concurrently across the entire cluster.

## Description

New configuration setting, e.g.:
`job_scheduler_concurrency_limits = event-processor-execution-v1:2,notification-execution-v1:2
`
When nothing is specified we default to the regular behavior, i.e. no limitations.

Note that it is OK for concurrency values to add up to more than the max number of worker threads; or even for a single value to exceed that. 

A meter `job_scheduler_executions_deny_rate` tracks the rate at which triggers are denied execution due to lock contention. 
Gauge `job_scheduler_executions_rescheduled` shows how many times jobs are being rescheduled before running.

## Motivation and Context

Provides granular control of resource allocation. See linked issue for more details.

## How Tested

Create a bunch of event definitions that run frequently. Observe execution deny rate is low.
Configure a low maximum concurrency level and restart server. Observe execution deny rate metric increases as available threads are insufficient to handle query rate.
Granular testing is really only possible via the integration test.

### CODE REVIEW

File: misc/graylog.conf [hunk 1]
```diff
@@ -746,3 +746,17 @@ mongodb_max_connections = 1000
 # Maximum value that can be set for an event limit.
 # Default: 1000
 #event_definition_max_event_limit = 1000
+
+# Optional limits on scheduling concurrency by job type. No more than the specified number of worker
+# threads will be executing jobs of the specified type across the entire cluster.
+# Default: no limitation
+# Note: Monitor job queue metrics to avoid excessive backlog of unprocessed jobs when using this setting!
+# Available job types:
```
Comment (thll) [inline misc/graylog.conf:754]: If we advertise the feature here, it should also work in Graylog open, so the code for parsing it needs to be moved to graylog2-server.

We should not list job types from the enterprise plugin here, those won't be available when the enterprise plugin is not present. Enterprise specific job types should probably be listed in our documentation then.

File: graylog2-server/src/main/java/org/graylog2/cluster/lock/MongoLockService.java [hunk 2]
```diff
@@ -89,6 +93,18 @@ public Optional<Lock> lock(@Nonnull String resource) {
         return lock(resource, null);
     }
 
+    @Override
+    public Optional<Lock> lock(@Nonnull String resource, int maxConcurrency) {
+        Preconditions.checkArgument(!Strings.isNullOrEmpty(resource));
+        final long l = collection.countDocuments(eq(FIELD_RESOURCE, resource));
+        if (l < maxConcurrency) {
```
Comment (thll) [inline graylog2-server/src/main/java/org/graylog2/cluster/lock/MongoLockService.java:100]: There is a unique index on `FIELD_RESOURCE` in that collection, so l can never be > 1

File: graylog2-server/src/main/java/org/graylog2/cluster/lock/MongoLockService.java [hunk 3]
```diff
@@ -89,6 +93,18 @@ public Optional<Lock> lock(@Nonnull String resource) {
         return lock(resource, null);
     }
 
+    @Override
+    public Optional<Lock> lock(@Nonnull String resource, int maxConcurrency) {
+        Preconditions.checkArgument(!Strings.isNullOrEmpty(resource));
+        final long l = collection.countDocuments(eq(FIELD_RESOURCE, resource));
```
Comment (thll) [inline graylog2-server/src/main/java/org/graylog2/cluster/lock/MongoLockService.java:99]: Performing a query and then attempting to lock based on the result is prone to race conditions. We can't do that.

An approach that should work has been outlined [here](https://github.com/Graylog2/graylog-plugin-enterprise/issues/6313#issuecomment-1878646487).

File: graylog2-server/src/main/java/org/graylog/scheduler/JobExecutionEngine.java [hunk 4]
```diff
@@ -125,28 +138,52 @@ public boolean execute() {
...
+            try (RefreshingLockService refreshingLockService = refreshingLockServiceFactory.create()) {
+                try {
+                    refreshingLockService.acquireAndKeepLock(
+                            trigger.jobDefinitionType(), maxConcurrencyMap.get(trigger.jobDefinitionType()));
+                    handleTrigger(trigger);
+                } catch (AlreadyLockedException e) {
+                    // Avoid endlessly retrying the same trigger (e.g. when concurrency limit is too low)
+                    // TODO: Consider smarter backoff strategy, e.g. counting number of consecutive failed attempts
+                    final DateTime nextTime = DateTime.now(DateTimeZone.UTC).plusSeconds(BACKOFF_FIVE_SECONDS);
+                    jobTriggerService.releaseTrigger(trigger, JobTriggerUpdate.withNextTime(trigger.nextTime()));
```
Comment (thll) [inline graylog2-server/src/main/java/org/graylog/scheduler/JobExecutionEngine.java:176]:
```suggestion
                    jobTriggerService.releaseTrigger(trigger, JobTriggerUpdate.withNextTime(nextTime));
```

**Review by thll — CHANGES_REQUESTED** [review]
How has this been tested? For me, it's not working as intended.

1. Regardless of the concurrency limit configured for a job type, the limit will always be effectively `1`.
2. If jobs with a concurrency limit are being rescheduled because the limit doesn't allow them being executed, they will eventually choke execution of other jobs.

I've added some comments inline but I think we should get the behaviour fixed before continuing with an in-depth review. Also, I think we will still have a problem with potential job starvation when the concurrency limit for a certain job type is exceeded and jobs need to be rescheduled.

I suggest we should find a solution for that problem first before investing more energy into the current approach.

I've created a pseudo-integration test that can be used as a starting point for verifying runtime properties of the job scheduler because I found it hard and cumbersome to test this manually. Maybe you'll find it useful.

I'll just add it here:
```
package org.graylog.scheduler;

import com.codahale.metrics.MetricRegistry;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonTypeName;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.jsontype.NamedType;
import com.github.joschi.jadconfig.util.Duration;
import com.google.common.util.concurrent.ThreadFactoryBuilder;
import com.google.common.util.concurrent.Uninterruptibles;
import org.bson.types.ObjectId;
import org.graylog.scheduler.capabilities.SchedulerCapabilitiesService;
import org.graylog.scheduler.clock.JobSchedulerClock;
import org.graylog.scheduler.clock.JobSchedulerSystemClock;
import org.graylog.scheduler.eventbus.JobSchedulerEventBus;
import org.graylog.scheduler.schedule.OnceJobSchedule;
import org.graylog.scheduler.worker.JobWorkerPool;
import org.graylog.testing.mongodb.MongoDBExtension;
import org.graylog.testing.mongodb.MongoDBTestService;
import org.graylog2.bindings.providers.CommonMongoJackObjectMapperProvider;
import org.graylog2.bindings.providers.MongoJackObjectMapperProvider;
import org.graylog2.cluster.lock.LockService;
import org.graylog2.cluster.lock.MongoLockService;
import org.graylog2.database.MongoCollections;
import org.graylog2.plugin.ServerStatus;
import org.graylog2.plugin.system.NodeId;
import org.graylog2.shared.bindings.providers.ObjectMapperProvider;
import org.graylog2.system.shutdown.GracefulShutdownService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.stream.IntStream;

import static org.assertj.core.api.Assertions.assertThat;

@ExtendWith(MongoDBExtension.class)
@ExtendWith(MockitoExtension.class)
@Timeout(10)
class JobSchedulerServiceIT {

    @Mock
    private ServerStatus serverStatus;

    @Mock
    private NodeId nodeId;

    @Mock
    SchedulerCapabilitiesService schedulerCapabilitiesService;

    @Mock
    private GracefulShutdownService gracefulShutdownService;

    private DBJobTriggerService jobTriggerService;
    private DBCustomJobDefinitionService customJobDefinitionService;
    private JobSchedulerService jobSchedulerService;
    private JobSchedulerClock clock;

    // intentionally mutable so that individual tests can add their factories
    private final Map<String, Job.Factory> jobFactories = new HashMap<>();

    @BeforeEach
    void setUp(MongoDBTestService mongoDBTestService) throws Exception {

        final ObjectMapper objectMapper = new ObjectMapperProvider(getClass().getClassLoader(), Set.of(
                new NamedType(UnlimitedJob.class, UnlimitedJob.TYPE_NAME),
                new NamedType(LimitedJobA.class, LimitedJobA.TYPE_NAME),
                new NamedType(OnceJobSchedule.class, OnceJobSchedule.TYPE_NAME)
        )).get();

        clock = new JobSchedulerSystemClock();

        final MetricRegistry metricRegistry = new MetricRegistry();
        final JobSchedulerEventBus eventBus = new JobSchedulerEventBus("job-scheduler", metricRegistry);
        final JobSchedulerConfig schedulerConfig = new TestSchedulerConfig();
        final Duration lockExpirationDuration = Duration.seconds(10);

        customJobDefinitionService = new DBCustomJobDefinitionService(
                new MongoCollections(new CommonMongoJackObjectMapperProvider(() -> objectMapper),
                        mongoDBTestService.mongoConnection()));

        jobTriggerService = new DBJobTriggerService(
                mongoDBTestService.mongoConnection(),
                new MongoJackObjectMapperProvider(objectMapper),
                nodeId,
                clock,
                schedulerCapabilitiesService,
                lockExpirationDuration
        );

        final DBJobDefinitionService jobDefinitionService = new DBJobDefinitionService(
                mongoDBTestService.mongoConnection(), new MongoJackObjectMapperProvider(objectMapper));
        final JobScheduleStrategies scheduleStrategies = new JobScheduleStrategies(clock);

        final JobTriggerUpdates.Factory jobTriggerUpdatesFactory = trigger -> new JobTriggerUpdates(
                clock,
                scheduleStrategies,
                trigger);

        final java.time.Duration leaderElectionLockTTL = java.time.Duration.ofSeconds(10);
        final LockService lockService = new MongoLockService(nodeId, mongoDBTestService.mongoConnection(), leaderElectionLockTTL);

        final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(30,
                new ThreadFactoryBuilder()
                        .setNameFormat("scheduled-daemon-%d")
                        .setDaemon(true)
                        .build()
        );

        final RefreshingLockServiceFactory refreshingLockServiceFactory = new RefreshingLockServiceFactory(
                lockService,
                scheduler,
                leaderElectionLockTTL);

        final JobExecutionEngine.Factory executionEnginFactory = workerPool ->
                new JobExecutionEngine(
                        jobTriggerService,
                        jobDefinitionService,
                        eventBus,
                        scheduleStrategies,
                        jobTriggerUpdatesFactory,
                        refreshingLockServiceFactory,
                        jobFactories,
                        workerPool,
                        schedulerConfig,
                        metricRegistry);

        final JobWorkerPool.Factory workerPoolFactory = (name, poolSize, shutdownCallback) ->
                new JobWorkerPool(name, poolSize, shutdownCallback, gracefulShutdownService, metricRegistry);

        final Duration loopSleepDuration = Duration.milliseconds(100);

        jobSchedulerService = new JobSchedulerService(executionEnginFactory, workerPoolFactory, schedulerConfig, clock, eventBus, serverStatus, loopSleepDuration);
    }

    @Test
    void testMaxConcurrency() throws Exception {
        final CountDownLatch outstandingJobs = new CountDownLatch(100);
        final ConcurrentHashMap<String, Integer> concurrentExecutions = new ConcurrentHashMap<>();
        final ConcurrentHashMap<String, Integer> maxConcurrentExecutions = new ConcurrentHashMap<>();

        final Job job = ctx -> {
            final String jobType = ctx.trigger().jobDefinitionType();
            final Integer runningJobs = concurrentExecutions.compute(jobType, (k, v) -> v == null ? 1 : v + 1);
            maxConcurrentExecutions.compute(jobType, (k, v) -> Math.max(runningJobs, v == null ? 0 : v));
            Uninterruptibles.sleepUninterruptibly(10, TimeUnit.MILLISECONDS);
            concurrentExecutions.compute(jobType, (k, v) -> v == null ? 0 : v - 1);
            outstandingJobs.countDown();
            return JobTriggerUpdate.withoutNextTime();
        };

        jobFactories.put(UnlimitedJob.TYPE_NAME, jobDefinitionDto -> job);
        jobFactories.put(LimitedJobA.TYPE_NAME, jobDefinitionDto -> job);

        jobSchedulerService.startAsync().awaitRunning();
        try {
            createTriggers(50, customJobDefinitionService.findOrCreate(jobDefinitionDto(new UnlimitedJob())));
            createTriggers(50, customJobDefinitionService.findOrCreate(jobDefinitionDto(new LimitedJobA())));
            outstandingJobs.await();
        } finally {
            jobSchedulerService.stopAsync().awaitTerminated();
        }

        assertThat(maxConcurrentExecutions.get(UnlimitedJob.TYPE_NAME)).isEqualTo(10);
        assertThat(maxConcurrentExecutions.get(LimitedJobA.TYPE_NAME)).isEqualTo(3);
    }

    @Test
    void testLimitedJobsAreNotBlockingOtherJobs() throws Exception {
        final CountDownLatch outstandingJobs = new CountDownLatch(50);

        final Job job = ctx -> {
            outstandingJobs.countDown();
            return JobTriggerUpdate.withoutNextTime();
        };

        final Job blockingJob = ctx -> {
            try {
                outstandingJobs.await();
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            return JobTriggerUpdate.withoutNextTime();
        };

        jobFactories.put(UnlimitedJob.TYPE_NAME, jobDefinitionDto -> job);
        jobFactories.put(LimitedJobA.TYPE_NAME, jobDefinitionDto -> blockingJob);

        jobSchedulerService.startAsync().awaitRunning();
        try {
            createTriggers(50, customJobDefinitionService.findOrCreate(jobDefinitionDto(new LimitedJobA())));
            createTriggers(50, customJobDefinitionService.findOrCreate(jobDefinitionDto(new UnlimitedJob())));
            outstandingJobs.await();
        } finally {
            jobSchedulerService.stopAsync().awaitTerminated();
        }
    }

    private List<JobTriggerDto> createTriggers(int numberOfTriggers, JobDefinitionDto jobDefinition) {
        return IntStream.range(0, numberOfTriggers).mapToObj(i ->
                jobTriggerService.create(JobTriggerDto.builder()
                        .jobDefinitionType(jobDefinition.config().type())
                        .jobDefinitionId(jobDefinition.id())
                        .schedule(OnceJobSchedule.create())
                        .build())).toList();
    }

    private static class TestSchedulerConfig implements JobSchedulerConfig {
        @Override
        public boolean canExecute() {
            return true;
        }

        @Override
        public int numberOfWorkerThreads() {
            return 10;
        }

        @Override
        public Map<String, Integer> jobMaxConcurrency() {
            return Map.of(LimitedJobA.TYPE_NAME, 3);
        }
    }

    @JsonTypeName(UnlimitedJob.TYPE_NAME)
    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class UnlimitedJob implements JobDefinitionConfig {
        public static final String TYPE_NAME = "unlimited-job";

        @Override
        public String type() {
            return TYPE_NAME;
        }
    }

    @JsonTypeName(LimitedJobA.TYPE_NAME)
    @JsonIgnoreProperties(ignoreUnknown = true)
    public static class LimitedJobA implements JobDefinitionConfig {
        public static final String TYPE_NAME = "limited-job-a";

        @Override
        public String type() {
            return TYPE_NAME;
        }
    }

    private JobDefinitionDto jobDefinitionDto(JobDefinitionConfig jobConfig) {
        return JobDefinitionDto.builder()
                .id(new ObjectId().toHexString())
                .title(JobSchedulerServiceIT.class.getSimpleName() + " job definition")
                .description("")
                .config(jobConfig)
                .build();
    }
}

```

File: graylog2-server/src/main/java/org/graylog2/cluster/lock/MongoLockService.java (hunk 3 above)
Comment (patrickmann) [reply graylog2-server/src/main/java/org/graylog2/cluster/lock/MongoLockService.java]: Trying to be too clever and limiting the number of DB queries. Assuming that concurrency limits are not that common, and that the backoff duration between retries is reasonable, we can probably get away with the sequential polling approach.

File: graylog2-server/src/main/java/org/graylog/scheduler/JobExecutionEngine.java (hunk 4 above)
Comment (patrickmann) [reply graylog2-server/src/main/java/org/graylog/scheduler/JobExecutionEngine.java]: Fixed

File: graylog2-server/src/main/java/org/graylog2/cluster/lock/MongoLockService.java (hunk 2 above)
Comment (patrickmann) [reply graylog2-server/src/main/java/org/graylog2/cluster/lock/MongoLockService.java]: Missed that. See below for more discussion.

File: misc/graylog.conf (hunk 1 above)
Comment (patrickmann) [reply misc/graylog.conf:754]: Done - apart from the docs update.

**Review by thll — APPROVED** [review]

**Pull request merged**